### DISCUSSION TITLE
RFC: Fabric "Molecules" (aka composed components)

### DISCUSSION THREAD

**Discussion by pearofducks** [fabric-ds/issues#105] [Ideas]
# Summary

Fabric needs guidance on how molecules should be handled; including when/how atoms should be modified, what support is created for molecules, and how molecules should be provided.

## Preface

For clarity, this document/discussion will make use of the _language_ established in [Atomic Design](https://bradfrost.com/blog/post/atomic-web-design).

This is *not* suggesting that Fabric follows Atomic Design principles, but it is useful to refer to core Fabric components as "atoms", and a composition of those components as a "molecule".

# Motivation (by example)

The concrete example of the "Card group" molecule will be used to highlight motivations.

An example in FINN is visible [here](https://pearofducks.com/fabric/rfc-card-group.png) in the "Størrelse på varen" area.

## Modifying atoms

The Card component currently sets `rounded-4` which cannot be overriden by `rounded-0`.

This means an implementation either needs support from the component, or to eject to pure CSS and bash the TW `rounded-4` class.

## Supporting molecules

Because the borders overlap, special CSS is needed for `hover` behavior.

Effectively:
- On hover, set `z-index: 1`
- When selected, set `z-index: 2 !important` - (`!important` so that hover behavior doesn't lower the element)

A working example can be seen [here](https://fabric-vue-card.surge.sh/).

## Providing molecules

The TJT team is currently packaging this molecule for use in their apps as a [standalone component](https://github.schibsted.io/finn/tjt-web/tree/main/common/components/packages/OptionCard).

# Implementation

TBD

**Comment by pearofducks** [comment]
# Modifying atoms for molecules

## When and how should we modify atoms to suit molecules?

- Do we tend to remove behavior from the atom to suit a molecule, and then let the molecule guide the specific behavior?
- Do we tend to add behavior to the atom, to suit specific molecules?

My recommendation here is we tend toward the former; steering an atom toward a more generic implementation and enabling the molecule to decide the specifics. In this case, we're removing (read: providing an option to remove) the `rounded-4` rule as it causes problems for the molecule's customization, instead of providing options to set `rounded-t-4`, `rounded-0`, or `rounded-b-4` automatically.

My worry being the latter ends up with "fat atoms" that have too many behaviors/cases baked into them.

## Do we know when we definitely do or don't want to modify an atom to suit a molecule (usecase)?

This is not so clear to me, but perhaps UXUI could make the call on if the molecule/pattern seems to be a good abstraction that could see reuse?

**Reply by digitalsadhu** [reply]
Is there a third option to “set a default behavior and allow molecules to override the behavior as needed”? 
Eg. Css vars with fallbacks?

**Reply by pearofducks** [reply]
This is definitely an option, great point!

This does, however, add size to the core CSS file instead of the individual component (JS) - so that's a tradeoff we'll have to consider. Unless we had these in `style` rules inline I guess, but that feels like it could get messy.

**Comment by pearofducks** [comment]
# Support for molecules

## Do we provide additional support for molecules? Always? Or only for certain conditions?

For the example case, the additional CSS is rather trivial if the atom (Card) is modified to not set border-radius. So this molecule could live as a copy-pasta in our tech documentation. This would imply that each user of the molecule owns one of the below solutions for support.

## What form is best for styling support?

If a molecule is too involved for copy-pasta, or we decide it meets criteria for first-class ownership or support - how do we do that?

### Custom TW-classes

In the example case, we could create/suggest a `hover:z-1` class and (f.ex) a `important:z-2` class.

### Custom component classes

For the example this would look like `f-card.grouped:hover` for the `z-index: 1` rule and `f-card.grouped.selected` for the `z-index: 2 !important` rule.

### Bake these stylings into the molecule's JS (where possible)

One technically could `onmouseover` and `onmouseout` for the hover, and not set that when the Card is in a selected state so `!important` isn't needed.

## Where should CSS support live?

- Molecule-rules get baked into core
- New stylesheet for molecules
- Per-molecule stylesheet

**Comment by pearofducks** [comment]
# Providing molecules

Perhaps we establish some threshold when something should be a documented-molecule vs. provided as a component?

## As documented combination of atoms

We already do provide some molecules via documentation - [e.g. the slightly more trivial case of toggles inside of cards](https://react.fabric-ds.io/card#toggles-inside-of-cards). So there's precident for this.

## As first class components

We don't have any of these at the moment. If we did start doing them, where do they live? Inside the library for the atoms - e.g. `@fabric-ds/react`? Or in another module?

**Comment by digitalsadhu** [comment]
The concepts of atoms and molecules are useful metaphors. Easy to grasp etc. as long as the atoms are solid, I wouldn’t have an issue with building out molecules on top so long as UI/UX has been through their standardisation process and there’s consensus about these molecule type components.

**Comment by digitalsadhu** [comment]
My first instinct is that ideally our “atoms” should mirror built in html elements in that they come with their own default styling and behaviour but in most cases, this styling and behaviour can be overridden. In some tricky cases, there’s even additional features added to the element to support customisation. Eg. The dialog element includes a ::backdrop pseudoselector to make it possible to target an otherwise hard to reach area of the component.

**Reply by pearofducks** [reply]
I'm having trouble seeing what this means practically. What would this imply for existing Fabric components that we're treating as atoms today?

**Reply by digitalsadhu** [reply]
Me too 😁
Was just a thought that popped into my head after reading those links
